### PR TITLE
identify tool for 1D mesh + snap on vertices

### PR DESCRIPTION
**Pull request by vcloarec**
## Description

[Replace this with some text explaining the rationale and details about this pull request]

### CODE REVIEW

File: src/core/mesh/qgsmeshlayer.h [hunk 1]
```diff
@@ -289,6 +291,37 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
...
+      * and so if the layer has not been rendered previously
+      * (e.g. when used in a script) it returns empty point
+      *
+      * \param pointCenter the center point of the search area, in map coordinates
+      * \param searchRadius the radius of the search area in map unit
+      * \returns coordinate point of the vertex in map coordinates
+      *
+      * \since QGIS 3.14
+      */
+    QgsPointXY snapOnVertex( const QgsPointXY &pointCenter, double searchRadius ) const;
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.h:331]: This function is not used anywhere , please remove

File: src/core/mesh/qgsmeshlayer.h [hunk 2]
```diff
@@ -289,6 +291,37 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
...
+      * (e.g. when used in a script) it returns NaN value
+      *
+      * \param index dataset index specifying group and dataset to extract value from
+      * \param rectangle the area where to search the edge, closest edge from center point will be used
+      * \returns interpolated value at the projected point. Returns NaN values for values
+      * outside the mesh layer and in case triangular mesh was not previously used for rendering
+      *
+      * \since QGIS 3.14
+      */
+    QgsMeshDatasetValue dataset1DValue( const QgsMeshDatasetIndex &index, const  QgsRectangle &rectangle ) const;
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.h]: use `double searchRadius` (no default value) to be consistent with `datasetValue`. Create rectangle internally.

File: src/core/mesh/qgsmeshlayer.h [hunk 3]
```diff
@@ -289,6 +291,37 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
       */
     QgsMesh3dDataBlock dataset3dValue( const QgsMeshDatasetIndex &index, const QgsPointXY &point ) const;
 
+    /**
+      * Returns the value of 1D mesh dataset defined on edge that are in the rectangle and the closest of the center of this rectangle
+      *
+      * \note It uses previously cached and indexed triangular mesh
+      * and so if the layer has not been rendered previously
+      * (e.g. when used in a script) it returns NaN value
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.h:304]: add reference to  \see updateTriangularMesh

File: src/core/mesh/qgsmeshlayer.h [hunk 4]
```diff
@@ -289,6 +291,37 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
...
+      *
+      * \param index dataset index specifying group and dataset to extract value from
+      * \param rectangle the area where to search the edge, closest edge from center point will be used
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.h]: use searchRadius

File: src/core/mesh/qgsmeshlayer.h [hunk 5]
```diff
@@ -257,20 +257,22 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
...
       * Interpolates the value on the given point from given dataset.
+      * For edge, the returned value is on the projected position on the edge
       *
       * \note It uses previously cached and indexed triangular mesh
       * and so if the layer has not been rendered previously
       * (e.g. when used in a script) it returns NaN value
       *
       * \param index dataset index specifying group and dataset to extract value from
       * \param point point to query in map coordinates
+      * \param rectangle zone to query for edge element, in map coordinates
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.h]: parameter is searchRadius. Specify which units user should use

File: src/core/mesh/qgsmeshlayer.h [hunk 6]
```diff
@@ -257,20 +257,22 @@ class CORE_EXPORT QgsMeshLayer : public QgsMapLayer
 
     /**
       * Interpolates the value on the given point from given dataset.
+      * For edge, the returned value is on the projected position on the edge
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.h]: For 3D datasets, it uses \s dataset3dValue 
For 1D datasets, it uses \s dataset3dValue with `searchRadius`

File: src/core/mesh/qgsmeshlayer.cpp [hunk 7]
```diff
@@ -352,6 +352,122 @@ QgsMesh3dDataBlock QgsMeshLayer::dataset3dValue( const QgsMeshDatasetIndex &inde
...
+    double sqrMaxDistFromPoint = pow( rectangle.width() / 2, 2 );
+    QgsPointXY projectedPoint;
+    for ( const int edgeIndex : edgeIndexes )
+    {
+      const QgsMeshEdge &edge = mesh->edges().at( edgeIndex );
+      const QgsMeshVertex &vertex1 = mesh->vertices()[edge.first];
+      const QgsMeshVertex &vertex2 = mesh->vertices()[edge.second];
+      QgsPointXY projPoint;
+      double sqrDist = centerPoint.sqrDistToSegment( vertex1.x(), vertex1.y(), vertex2.x(), vertex2.y(), projPoint );
+      if ( sqrDist < sqrMaxDistFromPoint )
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp:376]: you should probably look for closest edge, this logic just takes (randomly) a last edge found by index?

File: src/core/mesh/qgsmeshlayer.cpp [hunk 8]
```diff
@@ -352,6 +352,122 @@ QgsMesh3dDataBlock QgsMeshLayer::dataset3dValue( const QgsMeshDatasetIndex &inde
...
+          const QgsPoint p1 = mesh->vertices()[v1], p2 = mesh->vertices()[v2];
+          const QgsMeshDatasetValue val1 = dataProvider()->datasetValue( index, v1 );
+          const QgsMeshDatasetValue val2 = dataProvider()->datasetValue( index, v2 );
+          double length = p1.distance( p2 );
+          const double x = val1.x() + p1.distance( projectedPoint.x(), projectedPoint.y() ) * ( val2.x() - val1.x() ) / length;
+          double y = std::numeric_limits<double>::quiet_NaN();
+          bool isVector = dataProvider()->datasetGroupMetadata( index ).isVector();
+          if ( isVector )
+            y = val1.y() + p1.distance( projectedPoint.x(), projectedPoint.y() ) * ( val2.y() - val1.y() ) / length;;
+
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp]: Don't we have such function exactly in the mesh utils or somewhere in renderer to generate this value?

File: src/core/mesh/qgsmeshlayer.cpp [hunk 9]
```diff
@@ -352,6 +352,122 @@ QgsMesh3dDataBlock QgsMeshLayer::dataset3dValue( const QgsMeshDatasetIndex &inde
...
+          const int v1 = edge.first, v2 = edge.second;
+          const QgsPoint p1 = mesh->vertices()[v1], p2 = mesh->vertices()[v2];
+          const QgsMeshDatasetValue val1 = dataProvider()->datasetValue( index, v1 );
+          const QgsMeshDatasetValue val2 = dataProvider()->datasetValue( index, v2 );
+          double length = p1.distance( p2 );
+          const double x = val1.x() + p1.distance( projectedPoint.x(), projectedPoint.y() ) * ( val2.x() - val1.x() ) / length;
+          double y = std::numeric_limits<double>::quiet_NaN();
+          bool isVector = dataProvider()->datasetGroupMetadata( index ).isVector();
+          if ( isVector )
+            y = val1.y() + p1.distance( projectedPoint.x(), projectedPoint.y() ) * ( val2.y() - val1.y() ) / length;;
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp]: p1.distance( projectedPoint.x(), projectedPoint.y() ) can be used from calculation of x?

File: src/core/mesh/qgsmeshlayer.cpp [hunk 10]
```diff
@@ -352,6 +352,122 @@ QgsMesh3dDataBlock QgsMeshLayer::dataset3dValue( const QgsMeshDatasetIndex &inde
...
+        default:
+          break;
+      }
+    }
+  }
+
+  return value;
+}
+
+QgsPointXY QgsMeshLayer::snapOnVertex( const QgsPointXY &pointCenter, double searchRadius ) const
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshlayer.cpp:416]: remove?

**Review by PeterPetrik — COMMENTED** [review]
feel free to create offical PR

File: src/core/mesh/qgsmeshlayer.h (hunk 1 above)
Comment (vcloarec) [reply src/core/mesh/qgsmeshlayer.h:331]: Its aim is to be used with crayfish for tracing a line along which will be pot on graph the dataset value

File: src/core/mesh/qgsmeshlayer.cpp (hunk 7 above)
Comment (vcloarec) [reply src/core/mesh/qgsmeshlayer.cpp:376]: of course, I've forgot `sqrMaxDistFromPoint=sqrDist;` in the if scope

File: src/core/mesh/qgsmeshlayer.h (hunk 1 above)
Comment (PeterPetrik) [reply src/core/mesh/qgsmeshlayer.h:331]: all plots in crayfish usually just use datasetValue (/1d/3d) directoly? maybe I do not understand which plot?

Comment (vcloarec) [reply src/core/mesh/qgsmeshlayer.h:331]: The idea is to make longitudinal plot for 1D with the same principle that cross section plot. But to allow the line to fit on the 1D network, the user could click near vertices and the line vertices for plot will fir with network vertices.

Comment (PeterPetrik) [reply src/core/mesh/qgsmeshlayer.h:331]: ins't it possible to do directly in the python code?

File: src/core/mesh/qgsmeshlayer.cpp (hunk 8 above)
Comment (vcloarec) [reply src/core/mesh/qgsmeshlayer.cpp]: yes, I didn't see it.(https://github.com/qgis/QGIS/blob/3656b101ee26de3123beb6b92a87394700ea63bb/src/core/mesh/qgsmeshlayerutils.h#L135)... I extend it for QgsMeshDataValue.

**Pull request closed without merging**